tests/unittests/Foundation/NSProgressTests.mm

```diff
@@ -194,14 +191,110 @@ - (void)addParentTo:(NSProgress*)child {
+    ASSERT_NEAR(0.5f + (1.0f / 18.0f), [prog1 fractionCompleted], c_errorMargin);
+
+    // Go past-the-end on prog3, should add the 20 pending units to prog2
+    [prog3 setCompletedUnitCount:4];
+
+    ASSERT_EQ(4, [prog3 completedUnitCount]);
+    ASSERT_EQ(80, [prog2 completedUnitCount]);
+    ASSERT_EQ(0, [prog1 completedUnitCount]);
+
+    ASSERT_NEAR(4.0f / 3.0f, [prog3 fractionCompleted], c_errorMargin);
+    ASSERT_NEAR(80.0f / 120.0f, [prog2 fractionCompleted], c_errorMargin);
+    ASSERT_NEAR(80.0f / 120.0f, [prog1 fractionCompleted], c_errorMargin);
+
+    // Reset prog3 back to 0 (probably shouldn't ever be done in production)
+    [prog3 setCompletedUnitCount:0];
+
+    ASSERT_EQ(0, [prog3 completedUnitCount]);
+    ASSERT_EQ(80, [prog2 completedUnitCount]); // Stays the same
+    ASSERT_EQ(0, [prog1 completedUnitCount]);
+
+    ASSERT_EQ(0, [prog3 fractionCompleted]);
+    ASSERT_NEAR((60.0f / 120.0f) - (1.0f / 18.0f),
+                [prog2 fractionCompleted],
+                c_errorMargin); // Lowers by 4/3 * 20/120, despite completedUnitCount staying the same
+    ASSERT_NEAR((60.0f / 120.0f) - (1.0f / 18.0f), [prog1 fractionCompleted], c_errorMargin); // Lowers by 4/3 * 20/120
+
+    // Complete prog3 again (probably shouldn't ever be done in production)
+    [prog3 setCompletedUnitCount:3];
+
+    ASSERT_EQ(3, [prog3 completedUnitCount]);
+    ASSERT_EQ(100, [prog2 completedUnitCount]); // Increments again
+    ASSERT_EQ(0, [prog1 completedUnitCount]);
+
+    ASSERT_EQ(1, [prog3 fractionCompleted]);
+    ASSERT_NEAR((80.0f / 120.0f) - (1.0f / 18.0f), [prog2 fractionCompleted], c_errorMargin); // Increases by 20/120
+    ASSERT_NEAR((80.0f / 120.0f) - (1.0f / 18.0f), [prog1 fractionCompleted], c_errorMargin); // Increases by 20/120
+
+    // Change prog3's total unit count to 9
+    [prog3 setTotalUnitCount:9];
+
+    ASSERT_EQ(3, [prog3 completedUnitCount]);
+    ASSERT_EQ(100, [prog2 completedUnitCount]);
+    ASSERT_EQ(0, [prog1 completedUnitCount]);
+
+    ASSERT_NEAR(1.0f / 3.0f, [prog3 fractionCompleted], c_errorMargin);
+    ASSERT_NEAR(0.5f, [prog2 fractionCompleted], c_errorMargin); // lower by 20/120, increase by 1/3 * 20/120
+    ASSERT_NEAR(0.5f, [prog1 fractionCompleted], c_errorMargin); // lower by 20/120, increase by 1/3 * 20/120
+
+    // Reset prog2 to 0 (probably shouldn't ever be done in production)
+    [prog2 setCompletedUnitCount:0];
+
+    ASSERT_EQ(3, [prog3 completedUnitCount]);
+    ASSERT_EQ(0, [prog2 completedUnitCount]); // Hard resets without taking anything from prog3 into account
+    ASSERT_EQ(0, [prog1 completedUnitCount]);
+
+    ASSERT_NEAR(1.0f / 3.0f, [prog3 fractionCompleted], c_errorMargin);
+    ASSERT_NEAR(0, [prog2 fractionCompleted], c_errorMargin); // fractionCompleted floors at 0
+    ASSERT_NEAR(0, [prog1 fractionCompleted], c_errorMargin);
+
+    [prog2 resignCurrent];
+    [prog1 resignCurrent];
+}
+
```

Comment:
It'd be a pain, but it may be worth adding a test that changing a value on a child triggers a KVO notification on a parent -- specifically for the keys that this changeset fixes.